esmvaltool/diag_scripts/radiation_budget/radiation_budget.py

```diff
@@ -0,0 +1,480 @@
+"""Plot the global radiation budget."""
+
+# To run the doctests:
+# % cd ESMValTool/esmvaltool/
+# % python -m doctest diag_scripts/radiation_budget/radiation_budget.py
+import logging
+import os
+
+import iris
+import matplotlib.pyplot as plt
+import numpy as np
+from common import get_filenames, load_data
+
+from esmvaltool.diag_scripts.shared import (
+    group_metadata,
+    run_diagnostic,
+    save_figure,
+)
+
+CWD = os.path.abspath(os.path.dirname(__file__))
+STEPHENS_FILENAME = "Stephens_et_al_2012_obs_Energy_Budget.txt"
+DEMORY_FILENAME = "Demory_et_al_2014_obs_Energy_Budget.txt"
+
+
+def derive_additional_variables(cubes):
+    """Return input ``cubes`` with the additional cubes.
+
+    ``cubes`` must contain the variables specified in the recipe.
+
+    The additional cubes derived from the cubes in ``cubes`` are as follows:
+
+    * total_sw_cloud_forcing
+    * upward_sw_reflected_surface
+    * sw_reflected_clouds
+    * sw_absorbed_atm
+    * upward_lw_emitted_surface
+    * total_lw_cloud_forcing
+    * net_surface_radiation
+    * radiation_adsorbed_surface
+    * radiation_net_toa
+
+    Parameters
+    ----------
+    cubes : :class:`iris.cube.CubeList`
+        The cubes corresponding with the variables in the recipe.
+
+    Returns
+    -------
+    :class:`iris.cube.CubeList`
+        The input ``cubes`` with the additional cubes.
+    """
+    def _constraint(var_name):
+        return iris.Constraint(
+            cube_func=lambda cube: cube.var_name == var_name)
+
+    rss = cubes.extract_cube(_constraint("rss"))
+    rsdt = cubes.extract_cube(_constraint("rsdt"))
+    rsut = cubes.extract_cube(_constraint("rsut"))
+    rsutcs = cubes.extract_cube(_constraint("rsutcs"))
+    rsds = cubes.extract_cube(_constraint("rsds"))
+    rls = cubes.extract_cube(_constraint("rls"))
+    rlut = cubes.extract_cube(_constraint("rlut"))
+    rlutcs = cubes.extract_cube(_constraint("rlutcs"))
+    rlds = cubes.extract_cube(_constraint("rlds"))
+    hfss = cubes.extract_cube(_constraint("hfss"))
+    hfls = cubes.extract_cube(_constraint("hfls"))
```

Comment:
What kind of contingency do you have in mind? Currently, if the variable to be extracted isn't in the cubeList, [an exception is raised](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/cube.html#iris.cube.CubeList.extract_cube). Are you thinking that instead, the diagnostic should continue without a missing variable?

Reply:
@alistairsellar, how would you like to handle this? Do you think the diagnostic should be able to continue without missing variables?

Reply:
Personally I prefer things to fail if something is wrong, but I think this should be governed by a wider ESMValTool convention rather than making diagnostic-specific decisions. @valeriupredoi does ESMValTool have a general principle on whether it should fail or warn when inputs are missing?

Reply:
I have added a discussion about this: #2435. Can we address this in a future PR if changes are required?